pennylane_lightning/src/simulator/GateImplementationsLM.hpp

```diff
@@ -0,0 +1,1207 @@
+        const size_t rev_wire_shift = (static_cast<size_t>(1U) << rev_wire);
+        const size_t wire_parity = fillTrailingOnes(rev_wire);
+        const size_t wire_parity_inv = fillLeadingOnes(rev_wire + 1);
+
+        if (inverse) {
+            for (size_t k = 0; k < Util::exp2(num_qubits - 1); k++) {
+                const size_t i0 =
+                    ((k << 1U) & wire_parity_inv) | (wire_parity & k);
+                const size_t i1 = i0 | rev_wire_shift;
+                const std::complex<PrecisionT> v0 = arr[i0];
+                const std::complex<PrecisionT> v1 = arr[i1];
+                arr[i0] = std::conj(matrix[0B00]) * v0 +
+                          std::conj(matrix[0B10]) *
+                              v1; // NOLINT(readability-magic-numbers)
+                arr[i1] = std::conj(matrix[0B01]) * v0 +
+                          std::conj(matrix[0B11]) *
+                              v1; // NOLINT(readability-magic-numbers)
+            }
+        } else {
+            for (size_t k = 0; k < Util::exp2(num_qubits - 1); k++) {
+                const size_t i0 =
+                    ((k << 1U) & wire_parity_inv) | (wire_parity & k);
+                const size_t i1 = i0 | rev_wire_shift;
+                const std::complex<PrecisionT> v0 = arr[i0];
+                const std::complex<PrecisionT> v1 = arr[i1];
+                arr[i0] =
+                    matrix[0B00] * v0 +
+                    matrix[0B01] * v1; // NOLINT(readability-magic-numbers)
+                arr[i1] =
+                    matrix[0B10] * v0 +
+                    matrix[0B11] * v1; // NOLINT(readability-magic-numbers)
+            }
+        }
+    }
+    /**
+     * @brief Apply a two qubit gate to the statevector.
+     *
+     * @param arr Pointer to the statevector.
+     * @param num_qubits Number of qubits.
+     * @param matrix Perfect square matrix in row-major order.
+     * @param wires Wires the gate applies to.
+     * @param inverse Indicate whether inverse should be taken.
+     */
+    template <class PrecisionT>
+    static inline void
+    applyTwoQubitOp(std::complex<PrecisionT> *arr, size_t num_qubits,
+                    const std::complex<PrecisionT> *matrix,
+                    const std::vector<size_t> &wires, bool inverse = false) {
+        const size_t rev_wire0 = num_qubits - wires[1] - 1;
+        const size_t rev_wire1 = num_qubits - wires[0] - 1; // Controll qubit
+
+        const size_t rev_wire0_shift = static_cast<size_t>(1U) << rev_wire0;
+        const size_t rev_wire1_shift = static_cast<size_t>(1U) << rev_wire1;
+
+        const size_t rev_wire_min = std::min(rev_wire0, rev_wire1);
+        const size_t rev_wire_max = std::max(rev_wire0, rev_wire1);
+
+        const size_t parity_low = fillTrailingOnes(rev_wire_min);
+        const size_t parity_high = fillLeadingOnes(rev_wire_max + 1);
+        const size_t parity_middle =
+            fillLeadingOnes(rev_wire_min + 1) & fillTrailingOnes(rev_wire_max);
```

Comment:
I'd love to see some detail into how this indexing schema works, compared with the old system -- maybe we can have you give a short talk (or a few slides) on how this works from a memory access POV. I think having the wider PL team aware of such insights would be quite valuable.

Reply:
I would love to do so. I can prepare a short for one of the next PL team meetings.